[target-allocator] Improve TargetAllocator README (#1316) (#1472)
open-telemetry/opentelemetry-operator: cmd/otel-allocator/README.md

distribute targets of the PrometheusReceiver on all deployed Collector instances. The release version matches the
operator's most recent release as well.

> In essence, Prometheus Receiver configs are "stolen" and replaced with a http_sd_configs directive that points to the 

Reviewer: There are two things that happen here: one is the overriding of the service discovery in the collector's scrape config, and the other thing is the target allocator can add new scrape configs. I see them as two distinct actions because one can use the target allocator without the prometheus CR functionality.

> In essence, Prometheus Receiver configs are "stolen" and replaced with a http_sd_configs directive that points to the 

Author: I'll have a look at rewriting this section this weekend to clarify.

> In essence, Prometheus Receiver configs are "stolen" and replaced with a http_sd_configs directive that points to the 

Author: Updated

Allocator, these are then loadbalanced/sharded to the collectors. In addition, the TargetAllocator can discover targets via Prometheus CRs (currently ServiceMonitor,

> PodMonitor) which it presents on the `/jobs` endpoint.

Reviewer: The configuration is provided by the `/scrape_configs` endpoint, and the resultant jobs are on the `/jobs` endpoint

Author: Fixed

```yaml
  targetAllocator:
    enabled: true
    prometheusCR:
      enabled: true
```
The CRs can be filtered by labels as documented here: [api.md#opentelemetrycollectorspectargetallocatorprometheuscr](../../docs/api.md#opentelemetrycollectorspectargetallocatorprometheuscr)

The prometheus receiver in the deployed collector also has to know where the Allocator service exists. This is done by a
OpenTelemetry Collector operator specific config.
```yaml
  config: |
    receivers:
      prometheus:
        config:
          scrape_configs:
          - job_name: 'otel-collector'
        target_allocator:
          endpoint: http://my-targetallocator-service
          interval: 30s
          collector_id: collector-1
```
Upstream documentation here: [Prometheusreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver#opentelemetry-operator)

The TargetAllocator service is named based on the OpenTelemetryCollector CR name. `collector_id` should be unique per
collector instance, such as the pod name. This can be passed down to the pod with env vars and replaced in the config

> with `${ENV}` notation.

Author: I'm unsure of this detail. It seems to work, but I don't know Why. Perhaps someone with more understanding can confirm this is indeed correct?

Reviewer: You could replace this to be `${POD_NAME}` (it's automatically supplied by the operator [here](https://github.com/open-telemetry/opentelemetry-operator/blob/6d3da4219a9027aa49728d14036b84d47071b347/pkg/collector/container.go#L88))

Author: Updated

The CRDs naturally also has to exist for the Allocator to pick them up. The best place to get them is from
prometheus-operator: [Releases](https://github.com/prometheus-operator/prometheus-operator/releases). Only the CRDs for

> CRs that the Allocator watches for need to be deployed. They can be picked out from the bundle.yaml file.

Author: It's a bit of a shame that this is bundled upstream. Is there a way only link the parts we look at, while staying up to date with upstream?
Or just copy because the API is stable (v1)?

Reviewer: We mostly provide this ability for users migrating to opentelemetry from prometheus rather than someone starting out with the collector. For now, I'd recommend we may want to leave this out and open a follow up issue on deploying these CRs with the operator's CRs to make this easier at which point we can document this more effectively. What do you think @kristinapathak?

Author: Sounds good, resolving

### RBAC
The ServiceAccount that the TargetAllocator runs as, has to have access to the CRs. A role like this will provide that
access.
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: opentelemetry-targetallocator-cr-role
rules:
- apiGroups:
  - monitoring.coreos.com
  resources:
  - servicemonitors
  - podmonitors
  verbs:
  - '*'
```
In addition, the TargetAllocator needs the same permissions as a Prometheus instance would to find the matching targets
from the CR instances.
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: opentelemetry-targetallocator-role
rules:
- apiGroups: [""]
  resources:
  - nodes
  - nodes/metrics
  - services
  - endpoints

>   - pods

Author: Unsure if we actually need all of these

Reviewer: i think all of these are required if someone wants to use the `kubernetes_sd_config`

distribute targets of the PrometheusReceiver on all deployed Collector instances. The release version matches the
operator's most recent release as well.

In essence, Prometheus Receiver configs are "stolen" and replaced with a http_sd_configs directive that points to the 
Allocator, these are then loadbalanced/sharded to the collectors. In addition, the TargetAllocator can discover targets
via Prometheus CRs (currently ServiceMonitor, PodMonitor) which it presents as scrape configs and jobs on the
`/scrape_configs` and `/jobs` endpoints respectively.

# Usage
The `targetAllocator:` controls the TargetAllocator general properties. Full API spec can be found here: [api.md#opentelemetrycollectorspectargetallocator](../../docs/api.md#opentelemetrycollectorspectargetallocator)

> A basic example that deploys, but does nothing.

Reviewer 2: Maybe instead of 'does nothing' we could add a static config for `otel-collector`, to provide the simplest example of what allocator does?

Author: This is what the main README.md does - which demonstrates that the allocator places itself as a http_sd_config. What is confusing about that as a newcomer is that the shown scrape_config doesn't seem to do anything, and its really hard to understand which collector will scrape. This is tied into the allocation strategies (which I do not yet understand, so cannot really write a description of). The 'does nothing' case allows for less astonishment to me, until there is a more exact description/example of what happens.

Author: Updated to be a copy of the main readme.

```yaml
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: collector-with-ta
spec:
  mode: statefulset
  targetAllocator:
    enabled: true
  config: |
    receivers:
      prometheus:
        config:
          scrape_configs:
          - job_name: 'otel-collector'

    exporters:
      logging:

    service:
      pipelines:
        traces:

>           receivers: [prometheus]

Reviewer 3: `Prometheus` receiver should be the part of metrics pipeline. I think, same fix is needed on the project level Readme also: https://github.com/open-telemetry/opentelemetry-operator/blob/6d3da4219a9027aa49728d14036b84d47071b347/README.md?plain=1#L352

Author: Blind copy-paste error. Oops

Author: Fixed, and in main README too

          processors: []
          exporters: [logging]
```

## PrometheusCR specifics
TargetAllocator discovery of PrometheusCRs can be turned on with
```yaml
  targetAllocator:
    enabled: true
    prometheusCR:
      enabled: true

> ```

Reviewer 2: Maybe instead of a YAML snippet we could just specify this can be enabled with `prometheusCR.enabled=true` and link to docs for additional features?

Author: Sure, with a full path, I think that is OK.
The reason I did it this way is to show the the targetAllocator has to be enabled - something I can totally see someone leaving out.

Author: Changed to full path.

The CRs can be filtered by labels as documented here: [api.md#opentelemetrycollectorspectargetallocatorprometheuscr](../../docs/api.md#opentelemetrycollectorspectargetallocatorprometheuscr)

The prometheus receiver in the deployed collector also has to know where the Allocator service exists. This is done by a
OpenTelemetry Collector operator specific config.
```yaml
  config: |
    receivers:
      prometheus:
        config:
          scrape_configs:
          - job_name: 'otel-collector'
        target_allocator:
          endpoint: http://my-targetallocator-service
          interval: 30s
          collector_id: "${POD_NAME}"
```
Upstream documentation here: [Prometheusreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver#opentelemetry-operator)

The TargetAllocator service is named based on the OpenTelemetryCollector CR name. `collector_id` should be unique per
collector instance, such as the pod name. The `POD_NAME` environment variable is convenient since this is supplied
to collector instance pods by default.

> The CRDs naturally also has to exist for the Allocator to pick them up. The best place to get them is from

Reviewer 2: ```suggestion
The Prometheus CRDs also have to exist for the Allocator to pick them up. The best place to get them is from
```

Otherwise I'd agree with @jaronoff97 to perhaps leave this out for now and do it in follow up?

Author: I think it should be mentioned that this is not done by automagic (which is very common in the k8s world). It can be removed later in favour of a link to a proper doc. Accepting your change.
